napari/utils/colormaps/colormap.py

```diff
@@ -230,3 +235,62 @@ def default_color(self):
         return self.color_dict.get(None, (0, 0, 0, 0))
         # we provided here default color for backward compatibility
         # if someone is using DirectLabelColormap directly, not through Label layer
+
+
+def cast_labels_to_minimum_type_auto(
+    data: np.ndarray, num_colors: int
+) -> np.ndarray:
+    """Perform modulo operation based on number of colors
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Labels data to be casted.
+    num_colors : int
+        Number of unique colors in the data.
+
+    Returns
+    -------
+    np.ndarray
+        Casted labels data.
+    """
+    dtype = minimum_dtype_for_labels(num_colors + 1)
+
+    return _cast_labels_to_minimum_type_auto(data, num_colors, dtype)
+
+
+@numba.njit(parallel=True)
+def _cast_labels_to_minimum_type_auto(
+    data: np.ndarray, num_colors: int, dtype
+) -> np.ndarray:
+    result_array = np.zeros_like(data, dtype=dtype)
```

Comment:
Oof. This is a big step. However, I tried my best to optimise this in NumPy, avoiding unnecessary allocations, and got 10x worse speed...

```
In [14]: %timeit cm._cast_labels_to_minimum_type_auto(data, 50, np.uint8)
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
33 ms ± 2.19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [16]: %%timeit
    ...: np.add(np.mod(data, np.uint8(50), out=out, casting='unsafe'),
    ...:        np.uint8(1),
    ...:        where=(data >= 50) | (data < 0),
    ...:        out=out)
    ...:
388 ms ± 830 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Even if we optimise for unsigned ints, we are still close to 10x worse:

```
In [15]: %%timeit
    ...: np.add(np.mod(data, np.uint8(50), out=out, casting='unsafe'),
    ...:        np.uint8(1),
    ...:        where=(data >= 50),
    ...:        out=out)
    ...:
280 ms ± 458 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So, maybe we should try to bite the bullet. Alternatively, maybe we can dispatch to the above NumPy code if numba is not available, but recommend numba (and include it in [all]).

What do you think @Czaki?

Certainly, depending on numba would open up a *lot* of optimisations. But it feels like we are making a big decision under time pressure. 😅

Reply:
Ah, I just saw what you think. Yes I guess you are right that for 0.4.19 it is not a huge risk. But maybe since this is going into main, we make the dependency optional (as discussed above), and when backporting to 0.4.19 we tear out the compatibility code? Or vice versa, we can add non-numba option in a future PR, your call.

Reply:
I think that the best option will be to donate such code to some of our compiled dependencies (vispy, scikit?) 

Here I see a double allocation:
```
where=(data >= 50) | (data < 0),
```

I would rather open an issue for this problem than add much slower/memory consummimg code.